src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,820 @@
+// annotation / protobuf message for interface purposes.
+func (enc *Encoder) LastEncoded() (ts.Datapoint, error) {
+	if enc.numEncoded == 0 {
+		return ts.Datapoint{}, errNoEncodedDatapoints
+	}
+
+	return enc.lastEncodedDP, nil
+}
+
+// Len returns the length of the data stream.
+func (enc *Encoder) Len() int {
+	return enc.stream.Len()
+}
+
+func (enc *Encoder) encodeHeader() {
+	enc.encodeVarInt(currentEncodingSchemeVersion)
+	enc.encodeVarInt(uint64(enc.opts.ByteFieldDictionaryLRUSize()))
+	enc.encodeCustomSchemaTypes()
+}
+
+func (enc *Encoder) encodeCustomSchemaTypes() {
+	if len(enc.customFields) == 0 {
+		enc.encodeVarInt(0)
+		return
+	}
+
+	// Field numbers are 1-indexed so encoding the maximum field number
+	// at the beginning is equivalent to encoding the number of types
+	// we need to read after if we imagine that we're encoding a 1-indexed
+	// bitset where the position in the bitset encodes the field number (I.E
+	// the first value is the type for field number 1) and the values are
+	// the number of bits required to unique identify a custom type instead of
+	// just being a single bit (3 bits in the case of version 1 of the encoding
+	// scheme.)
+	maxFieldNum := enc.customFields[len(enc.customFields)-1].fieldNum
+	enc.encodeVarInt(uint64(maxFieldNum))
+
+	// Start at 1 because we're zero-indexed.
+	for i := 1; i <= maxFieldNum; i++ {
+		customTypeBits := uint64(cNotCustomEncoded)
+		for _, customField := range enc.customFields {
+			if customField.fieldNum == i {
+				customTypeBits = uint64(customField.fieldType)
+				break
+			}
+		}
+
+		enc.stream.WriteBits(
+			customTypeBits,
+			numBitsToEncodeCustomType)
+	}
+}
+
+func (enc *Encoder) encodeTimestamp(t time.Time, tu xtime.Unit) error {
+	if !enc.hasEncodedFirstSetOfCustomValues {
+		return enc.m3tszEncoder.WriteFirstTime(t, nil, tu)
+	}
+	return enc.m3tszEncoder.WriteNextTime(t, nil, tu)
+}
+
+// TODO: Add concept of hard/soft error and if there is a hard error
+// then the encoder cant be used anymore.
+func (enc *Encoder) encodeProto(m *dynamic.Message) error {
+	if len(m.GetUnknownFields()) > 0 {
+		// TODO(rartoul): Make this behavior configurable.
+		return errEncoderMessageHasUnknownFields
+	}
+
+	if err := enc.encodeCustomValues(m); err != nil {
+		return err
+	}
+	if err := enc.encodeProtoValues(m); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reset resets the encoder for reuse.
+func (enc *Encoder) Reset(
+	start time.Time,
+	capacity int,
+) {
+	enc.reset(start, capacity)
+}
+
+// SetSchema sets the encoders schema.
+func (enc *Encoder) SetSchema(schema *desc.MessageDescriptor) {
+	enc.resetSchema(schema)
+}
+
+func (enc *Encoder) reset(start time.Time, capacity int) {
+	// TODO: Probably don't want to make these both the same capacity.
+	// TODO: Won't this be a massive alloc on the m3tsz encoder side for no reason?
+	// probably need to be able to reset the encoder and also specify the OStream
```

Comment:
Aren't you sharing the same ostream?  This is strange.

Reply:
Yeah this might just be stale, will look into it.

Reply:
Yeah ok its a little gross, but actually all I need to do is call reset on the m3tsz encoder because that will reset the shared OStream. I'll add a comment but this might get better when I implemented your `EncoderStream` suggestion